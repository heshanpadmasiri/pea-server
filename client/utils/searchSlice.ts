import { createSlice } from '@reduxjs/toolkit';

export const searchSlice = createSlice({
    name: 'search',
    initialState: {
        query: <undefined | string>undefined,
    },
    reducers: {
        clearQuery: (state) => {
            state.query = undefined;
        },
        setQuery: (state, action) => {
            const newQuery = action.payload as string;
            state.query = newQuery;
        }
    }
})

export const { clearQuery, setQuery } = searchSlice.actions;
export default searchSlice.reducer;
