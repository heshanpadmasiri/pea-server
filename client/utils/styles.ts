import { StyleSheet } from 'react-native';
const styles = StyleSheet.create({
    safeArea: {
        flex: 5,
        paddingHorizontal: 10,
    },
    searchBar: {
        flexDirection: 'row',
        flex: 0.25,
        alignItems: 'center',
        justifyContent: 'space-between',
    },
    gallery: {
        flex: 1,
        backgroundColor: 'black'
    },
    grid: {
        flexDirection: 'row',
    },
    grid_image: {
        height: 300,
        width: 300,
        resizeMode:'contain',
    },
    tag_selector: {
        flex: 3,
    },
    switch_container: {
        flex: 1,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
    },
    thumbnailCard: {
        alignItems: 'center',
        backgroundColor: 'white',
        marginBottom: 10,
        height: 220,
    },
    thumbnailHeading: {
        flex: 1
    },
    thumbnail: {
        height: 200,
        width: 200
    },
    container: {
        flex: 1,
        justifyContent: 'center',
        backgroundColor: '#ecf0f1',
    },
    video: {
        alignSelf: 'center',
        width: 320,
        height: 200,
    },
    buttons: {
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'center',
    },
});


export default styles;
