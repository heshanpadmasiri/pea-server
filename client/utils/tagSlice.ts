import { createSlice } from "@reduxjs/toolkit";

export const tagSlice = createSlice({
    name: "tags",
    initialState: {
        selectedTags: <string[]>[],
    },
    reducers: {
        selectTag: (state, action) => {
            if (state.selectedTags.includes(action.payload)) {
                return;
            }
            state.selectedTags.push(action.payload);
        },
        unselectTag: (state, action) => {
            state.selectedTags = state.selectedTags.filter((tag) => tag !== action.payload);
        },
    }
});

export const { selectTag, unselectTag } = tagSlice.actions;

export default tagSlice.reducer;
