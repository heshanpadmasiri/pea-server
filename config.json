{
    "url": "https://c1a56103-7250-419c-8033-e51d4ebb6cb7-dev.e1-us-east-azure.choreoapis.dev/zlkb/redirection-service/1.0.0",
    "auth": "eyJraWQiOiJnYXRld2F5X2NlcnRpZmljYXRlX2FsaWFzIiwiYWxnIjoiUlMyNTYifQ.eyJzdWIiOiI4ZWU3YTA3OS01YjcxLTRkZjYtYWI5YS01MDNkNDk4ZjJhYTNAY2FyYm9uLnN1cGVyIiwiaXNzIjoiaHR0cHM6XC9cL3N0cy5jaG9yZW8uZGV2OjQ0M1wvb2F1dGgyXC90b2tlbiIsImtleXR5cGUiOiJQUk9EVUNUSU9OIiwic3Vic2NyaWJlZEFQSXMiOlt7InN1YnNjcmliZXJUZW5hbnREb21haW4iOm51bGwsIm5hbWUiOiJSRVNUIGFwaSIsImNvbnRleHQiOiJcL2MxYTU2MTAzLTcyNTAtNDE5Yy04MDMzLWU1MWQ0ZWJiNmNiN1wvemxrYlwvcmVkaXJlY3Rpb24tc2VydmljZVwvMS4wLjAiLCJwdWJsaXNoZXIiOiJjaG9yZW9fcHJvZF9hcGltX2FkbWluIiwidmVyc2lvbiI6IjEuMC4wIiwic3Vic2NyaXB0aW9uVGllciI6bnVsbH1dLCJleHAiOjE2NzY3NjYyOTQsInRva2VuX3R5cGUiOiJJbnRlcm5hbEtleSIsImlhdCI6MTY3NjcwNjI5NCwianRpIjoiYWU5NDQ0OGMtYTM1Ni00MmIwLWJkZDAtNWZmN2IxNmIxY2M0In0.jHVSVWlYdqIm4xhNCDgqfJKElVeWeuymTEMbNI8oMyGkN0MZ3Z4eXJe_iDm1QPfPMt5eneYANV1LikvCfk_hEYxrEQO4yAo4fRtz3MlmU__G7Njuqjqwfc-kqHJOgsh1nxXEw4BD7lUQ7PgzY-FzB90-WnroRWC_ZMVsM__ZkoEWslw2BUgrTgU8uvlARZKv-UY7kZBZQvRH5GPrqUxmH0mD6MKS_GenVu_TPQlldzSFmfZhJBegbgNONs8d3PooBcbF00EH4_YCaMo3NgmB3xCRMH-FWVEu1USm7pc5zS7GS9VKqdtTK_Pm379IwZSVU5i646SsGJYZkxJ2idA8WQaD4OOsivu4g7LxN6jQ7AEELHBk4iTKaGX7J7FwZyUu4rdjeRFEbHPntmkBou2cZMCnOVL7VikbHyLBAMUlsi9bP3bId2kQiFqqEBd9sXLkZh1SWhg6RNfIeG8oHKeGXXJDqn0XQpFL_rt0DGydShsv-9AgNwHUZGTt3Kuwb3epr7_wIaU7tu_LEEwrzXNmKCkdWwySK8N1xRT8QKNv7wt0l4yM1Cavsak1kw2cIPbkDXMmutEpucC5W1HFtWkBXuMsE9hpDLu8CNSL-zffFX7lKfGzUFXWh22p-pE5_nUYZDpQzWv4WaBeBrDq7SZwJFBVSSTXOyBM2UNa6QZ9_8A"
}
