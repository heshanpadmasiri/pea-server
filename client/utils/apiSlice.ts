import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import config from '../config.json';

export interface Metadata {
    name: string,
    id: string,
    ty: string,
    tags: string[],
}

export type Condition = {
    type: string,
    tags: string[]
}

export type QueryResult = {
    isLoading: boolean,
    isError: boolean,
    isSuccess: boolean,
    error: any,
    data: Metadata[]
}

export function fileContentUrl(file: Metadata): string {
    return config.SERVER_URL + '/content/' + file.id;
}

export const apiSlice = createApi({
    reducerPath: 'api',
    baseQuery: fetchBaseQuery({ baseUrl: config.SERVER_URL }),
    endpoints: (builder) => ({
        getTags: builder.query<string[], void>({
            query: () => '/tags'
        }),
        getFiles: builder.query<Metadata[], void>({
            query: () => '/files'
        }),
        getFilesByType: builder.query<Metadata[], string>({
            query: (type) => `/files/${type}`
        }),
        getFilesByCondition: builder.query<Metadata[], Condition>({
            query: ({ type, tags }) => ({
                url: '/query',
                method: 'POST',
                body: {
                    data: { tags, ty: type }
                }
            })
        })
    })
});

export const { useGetTagsQuery, useGetFilesQuery, useGetFilesByTypeQuery, useGetFilesByConditionQuery } = apiSlice;
