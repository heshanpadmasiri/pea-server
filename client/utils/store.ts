import { configureStore } from '@reduxjs/toolkit';
import tagReducer from './tagSlice';
import slideShowReducer from './slideShowSlice';
import { apiSlice } from './apiSlice';

const store = configureStore({
    reducer: {
        tags: tagReducer,
        slideShow: slideShowReducer,
        [apiSlice.reducerPath]: apiSlice.reducer
    },
    middleware: getDefaultMiddleware => getDefaultMiddleware(
        { serializableCheck: false }
    ).concat(apiSlice.middleware)
});

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch

export default store;
