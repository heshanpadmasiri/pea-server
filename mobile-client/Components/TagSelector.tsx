import { useEffect, useState } from "react";
import { View, Text } from "react-native";
import { getTags } from "../utils/services";
import { get_data_and_update_state } from "../utils/states";
import styles from "../utils/styles";

export type TagSelectorProps = {
    updateSelectedTags: (tags: string[]) => void;
    selectedTags: string[];
}

const TagSelector = (props: TagSelectorProps) => {
    const [isLoading, setIsLoading] = useState(true);
    const [initialized, setInitialized] = useState(false);
    const [isError, setIsError] = useState(false);
    const [selectedTags, setSelectedTags] = useState<string[]>(props.selectedTags);
    const [tags, setTags] = useState<string[]>([]);

    useEffect(() => {
        if(!initialized) {
            get_data_and_update_state<string>(getTags, setTags, setIsLoading, setIsError);
            setInitialized(true);
        }
    });


    if (isLoading) {
        return (
            <View style={styles.container}>
                <Text>Loading...</Text>
            </View>
        );
    }
    else if (isError) {
        return (
            <View style={styles.container}>
                <Text>Error!</Text>
            </View>
        );
    }
    else {
        return (<Text>Tag selector</Text>);
    }
}

export default TagSelector
