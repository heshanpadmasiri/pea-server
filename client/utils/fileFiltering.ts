import { useSelector } from 'react-redux';
import { QueryResult, useGetFilesQuery } from './apiSlice';
import { RootState } from './store';
import { useGetFilesByConditionQuery, useGetFilesByTypeQuery } from './apiSlice';

export function getFilesByType(type: string): QueryResult {
    const selectedTags = useSelector((state: RootState) => state.tags.selectedTags);
    let queryResult: QueryResult;
    if (selectedTags.length > 0) {
        queryResult = useGetFilesByConditionQuery({ type, tags: selectedTags }) as QueryResult;
    }
    else {
        queryResult = useGetFilesByTypeQuery(type) as QueryResult;
    }
    return filterBySearchQuery(queryResult);
}

export function getFilesByTypes(types: string[]): QueryResult[] {
    return types.map((type) => getFilesByType(type))
}

export function getAllFiles(): QueryResult {
    const selectedTags = useSelector((state: RootState) => state.tags.selectedTags);
    let queryResult: QueryResult;
    if (selectedTags.length > 0) {
        queryResult = useGetFilesByConditionQuery({ type: '', tags: selectedTags }) as QueryResult;
    }
    else {
        queryResult = useGetFilesQuery() as QueryResult;
    }
    return filterBySearchQuery(queryResult);
}

function filterBySearchQuery(queryResult: QueryResult): QueryResult {
    const query = useSelector((state: RootState) => state.search.query);
    if (query === undefined) {
        return queryResult;
    }
    if (queryResult.isError || queryResult.isLoading) {
        return queryResult;
    }
    const data = queryResult.data.filter((file) => file.name.toLowerCase().includes(query.toLowerCase()));
    return { ...queryResult, data };
}
