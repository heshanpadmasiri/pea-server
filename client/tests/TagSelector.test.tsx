// import { fetch } from 'cross-fetch';
// const { Headers, Request, Response } = fetch;

import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { render, screen } from '@testing-library/react-native'
import config from '../config.json';
import { setupStore } from '../utils/store';
import { Provider } from 'react-redux';
import TagSelector from '../Components/TagSelector';

// No idea why I need these but some how even after giving a fetchFn I am getting errors without these
// global.Headers = Headers;
// global.Request = Request;
// global.Response = Response;

const BASE_URL = config.SERVER_URL;
export const handlers = [
    rest.get(`${BASE_URL}/tags`, (_req, res, ctx) => {
        return res(ctx.json(['tag 1', 'tag 2']), ctx.status(200));
    })
]

const server = setupServer(...handlers);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
const delay = (ms: number) => new Promise(res => setTimeout(res, ms));
test('renders component', async () => {
    const store = setupStore();
    render(
        <Provider store={store}>
            <TagSelector />
        </Provider>
    );
    // This line is not working in github and due to some reason hanging in both local and github why?
    await screen.findByTestId('recieved-tag');
    // delay(1000); 
    expect(screen.toJSON()).toMatchSnapshot()
});
