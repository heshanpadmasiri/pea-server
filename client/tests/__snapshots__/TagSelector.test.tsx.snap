// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component 1`] = `
<View
  style={
    {
      "backgroundColor": "#ecf0f1",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    style={
      {
        "flex": 1,
      }
    }
    testID="recieved-tag"
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text>
          tag 1
        </Text>
        <RCTSwitch
          accessibilityRole="switch"
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onTintColor="#81b0ff"
          style={
            [
              {
                "height": 31,
                "width": 51,
              },
              {
                "backgroundColor": "#3e3e3e",
                "borderRadius": 16,
              },
            ]
          }
          thumbTintColor="#f4f3f4"
          tintColor="#767577"
          value={false}
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text>
          tag 2
        </Text>
        <RCTSwitch
          accessibilityRole="switch"
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onTintColor="#81b0ff"
          style={
            [
              {
                "height": 31,
                "width": 51,
              },
              {
                "backgroundColor": "#3e3e3e",
                "borderRadius": 16,
              },
            ]
          }
          thumbTintColor="#f4f3f4"
          tintColor="#767577"
          value={false}
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
